pycqed/analysis_v2/residual_ZZ_analysis.py

```diff
@@ -0,0 +1,192 @@
+            self.run_analysis()
+        
+    def extract_data(self):
+        self.raw_data_dict = {}
+        self.timestamps = a_tools.get_timestamps_in_range(
+            self.t_start, self.t_stop,
+            label=self.labels)
+        for ts in self.timestamps:
+            data_fp = get_datafilepath_from_timestamp(ts)
+            param_spec = {
+                'data': ('Experimental Data/Data', 'dset'),
+                'target_qubit':  ('Experimental Data/Experimental Metadata', 'attr:target_qubit'),
+                'spectator_qubits':  ('Experimental Data/Experimental Metadata', 'attr:spectator_qubits'),
+                'spectator_state':  ('Experimental Data/Experimental Metadata', 'attr:spectator_state')}
+            self.raw_data_dict[ts] = h5d.extract_pars_from_datafile(
+                data_fp, param_spec)
+        
+        # Parts added to be compatible with base analysis data requirements
+        self.raw_data_dict['timestamps'] = self.timestamps
+        self.raw_data_dict['folder'] = os.path.split(data_fp)[0]
+        
+    def process_data(self):
+            self.proc_data_dict = {}
+            residual_ZZ_matrix = np.zeros((len(self.qubits),len(self.qubits)))
+            for ts in self.timestamps:
+                self.timestamp = ts
+                target_qubit = self.raw_data_dict[ts]['target_qubit']
+                spectators = self.raw_data_dict[ts]['spectator_qubits'] 
+                state = self.raw_data_dict[ts]['spectator_state'] 
+                spectator_list = spectators.strip('[').strip(']').split(', ')
+                for i, bit in enumerate(state):
+                    if int(bit) == 1:
+                        active_spectator = spectator_list[i]
+                        active_spectator = active_spectator[1:-1]
+                x, y = self.raw_data_dict[ts]['data'][:-2,0], self.raw_data_dict[ts]['data'][:-2,1]
+                fit_res = fit_residual_coupling_oscillation(x, y)
+                frequency = fit_res.best_values['frequency']
+                if y[1]-y[0] < 0:
+                    frequency = -1*frequency
+                if np.std(y)/np.mean(y) < 0.1:
+                    frequency = 0
+                residual_ZZ_matrix[self.qubits.index(target_qubit), self.qubits.index(active_spectator)] = frequency
+                print('{}:Setting {},{} to {}'.format(ts, target_qubit, active_spectator, frequency))
+            for i in range(len(self.qubits)):
+                residual_ZZ_matrix[i,i] = np.nan
+            self.proc_data_dict['quantities_of_interest'] = {
+                'matrix': residual_ZZ_matrix,
+                'qubit_labels': self.qubits}
+    def prepare_plots(self):
+        self.plot_dicts['residual_matrix'] = {
+            'plotfn': plot_crosstalk_matrix,
+            'crosstalk_matrix': self.proc_data_dict['quantities_of_interest']['matrix'],
+            'qubit_labels': self.qubits
+        }
+
+def fit_residual_coupling_oscillation(x, y, **kw):
```

Comment:
So you are fitting residual coupling oscillations. 
Why are we not plotting this? Or creating a range of plots. 
I think those underlying plots are extremely important. 
Whenever we look at the matrix you produce and see something weird, the first thing to do is a sanity check, does this make sense? Right now that is virtually impossible.

Reply:
Related to this, one can now not run this residual ZZ analysis on a single coupling, only on a grid containing multiple ones.